### PR TITLE
rt: use optional non-zero value for task `owner_id`

### PR DESCRIPTION
**Pull request by hds**
We switch to using a `NonZeroU64` for the `id` field for `OwnedTasks`
and `LocalOwnedTasks` lists. This allows the task header to contain an
`Option<NonZeroU64>` instead of a `u64` with a special meaning for 0.

The size in memory will be the same thanks to Rust's niche optimization,
but this solution is clearer in its intent.

### CODE REVIEW

File: tokio/src/runtime/task/list.rs [hunk 1]
```diff
@@ -170,11 +171,7 @@ impl<S: 'static> OwnedTasks<S> {
     }
 
     pub(crate) fn remove(&self, task: &Task<S>) -> Option<Task<S>> {
-        let task_id = task.header().get_owner_id();
-        if task_id == 0 {
-            // The task is unowned.
-            return None;
-        }
+        let task_id = task.header().get_owner_id()?;
```
Comment (Darksonn) [inline tokio/src/runtime/task/list.rs:174]: I'd like at least a comment here to point out what's going on.

File: tokio/src/runtime/task/core.rs [hunk 2]
```diff
@@ -394,13 +395,13 @@ impl Header {
     }
 
     // safety: The caller must guarantee exclusive access to this field, and
-    // must ensure that the id is either 0 or the id of the OwnedTasks
+    // must ensure that the id is either `None` or the id of the OwnedTasks
     // containing this task.
-    pub(super) unsafe fn set_owner_id(&self, owner: u64) {
+    pub(super) unsafe fn set_owner_id(&self, owner: Option<NonZeroU64>) {
         self.owner_id.with_mut(|ptr| *ptr = owner);
     }
```
Comment (Darksonn) [inline tokio/src/runtime/task/core.rs:402]: This is never called with `None`, so you might as well just take a `NonZeroU64` as argument?

File: tokio/src/runtime/task/core.rs [hunk 3]
```diff
@@ -162,7 +163,7 @@ pub(crate) struct Header {
     /// Table of function pointers for executing actions on the task.
     pub(super) vtable: &'static Vtable,
 
-    /// This integer contains the id of the OwnedTasks or LocalOwnedTasks that
+    /// This non-zero integer contains the id of the OwnedTasks or LocalOwnedTasks that
     /// this task is stored in. If the task is not in any list, should be the
     /// id of the list that it was previously in, or zero if it has never been
     /// in any list.
```
Comment (Darksonn) [inline tokio/src/runtime/task/core.rs:169]:
```suggestion
    /// This integer contains the id of the OwnedTasks or LocalOwnedTasks that
    /// this task is stored in. If the task is not in any list, should be the
    /// id of the list that it was previously in, or `None` if it has never been
    /// in any list.
```

File: tokio/src/runtime/task/core.rs (hunk 2 above)
Comment (hds) [reply tokio/src/runtime/task/core.rs]: Done.

File: tokio/src/runtime/task/list.rs (hunk 1 above)
Comment (hds) [reply tokio/src/runtime/task/list.rs]: Done.

File: tokio/src/runtime/task/list.rs [hunk 4]
```diff
@@ -257,11 +254,7 @@ impl<S: 'static> LocalOwnedTasks<S> {
...
```
[the remark this answers is not in the document]
Comment (hds) [reply tokio/src/runtime/task/list.rs]: Done.

**Review by Darksonn — APPROVED** [review]

**Pull request merged**